src/zcash/Note.cpp

```diff
@@ -38,6 +40,59 @@ uint256 SproutNote::nullifier(const SproutSpendingKey& a_sk) const {
     return PRF_nf(a_sk, rho);
 }
 
+// Create a note based on a given spending key, with random r and value.  Useful for testing.
+SaplingNote SaplingNote::random(const SaplingSpendingKey& sk) {
+    auto addr = sk.default_address().get();
+    auto d = addr.d;
+    auto pk_d = addr.pk_d;
+    uint256 r;
+    librustzcash_sapling_generate_r(r.begin());
+    auto value = GetRand(10000);
+    return SaplingNote(d, pk_d, value, r);
+}
+
+// Call librustzcash to compute the commitment
+uint256 SaplingNote::cm() const {
```

Comment:
I would rather return an optional from this.

Reply:
I would return an optional from librustzcash if I could cross FFI boundaries with such types. :) In the same way that we return an optional from APIs involving diversifiers. I don't want anyone forgetting that commitments cannot always be computed; they could in the old code but they can't now.

If using an exception here makes something else easier then you could dismiss my review. Everything else looked fine.

Reply:
For future: the librustzcash function could return an int value, instead of bool, where 0=ok, and anything else is some kind of errcode which is defined in the header file.

Reply:
Pushed a commit to return optional.